test/e2e/fqdn_dns_cache_test.go

```diff
@@ -0,0 +1,391 @@
+	agentChanges := func(config *agentconfig.AgentConfig) {
+		config.DNSServerOverride = dnsServiceIP
+	}
+	return data.mutateAntreaConfigMap(nil, agentChanges, false, true)
+}
+
+// createCustomDnsService creates the custom DNS service and returns its ClusterIP as ipv4 address.
+func createCustomDnsService(data *TestData) (string, error) {
+	const dnsPort = 53
+	customDnsService, err := data.CreateServiceWithAnnotations("custom-dns-service", data.testNamespace, dnsPort,
+		dnsPort, corev1.ProtocolUDP, map[string]string{"app": "custom-dns"}, false,
+		false, corev1.ServiceTypeClusterIP, ptr.To[corev1.IPFamily](corev1.IPv4Protocol), map[string]string{})
+	if err != nil {
+		return "", err
+	}
+	return customDnsService.Spec.ClusterIP, nil
+
+}
+
+// createDnsConfig generates a DNS configuration for the specified IP address and domain name.
+func createDnsConfig(t *testing.T, ipv4Address, domainName string) map[string]string {
+	const coreFileTemplate = `lfx.test:53 {
+        errors
+        log
+        health
+        hosts {
+            {{ if .IPv4 }}{{ .IPv4 }} {{ $.FQDN }}{{ end }}
+            no_reverse
+            pods verified
+            ttl 10
+        }
+        loop
+        reload
+    }`
+
+	generateConfigData := func() (string, error) {
+		data := struct {
+			IPv4 string
+			FQDN string
+		}{
+			IPv4: ipv4Address,
+			FQDN: domainName,
+		}
+
+		tmpl, err := template.New("configMapData").Parse(coreFileTemplate)
+		if err != nil {
+			return "", err
+		}
+		var output bytes.Buffer
+		err = tmpl.Execute(&output, data)
+		if err != nil {
+			return "", err
+		}
+		return strings.TrimSpace(output.String()), nil
+	}
+
+	configMapData, err := generateConfigData()
+	require.NoError(t, err, "error processing configData template for DNS, ", err)
+	configData := map[string]string{
+		"Corefile": configMapData,
+	}
+
+	return configData
+}
+
+// createDnsPod creates the CoreDNS pod configured to use the custom DNS ConfigMap.
+func createDnsPod(t *testing.T, data *TestData) {
+	const customDnsImage = "coredns/coredns:1.11.3"
+
+	volume := []corev1.Volume{
+		{
+			Name: "config-volume",
+			VolumeSource: corev1.VolumeSource{
+				ConfigMap: &corev1.ConfigMapVolumeSource{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: "custom-dns-config",
+					},
+					Items: []corev1.KeyToPath{
+						{
+							Key:  "Corefile",
+							Path: "Corefile",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	volumeMount := []corev1.VolumeMount{
+		{
+			Name:      "config-volume",
+			MountPath: "/etc/coredns",
+		},
+	}
+
+	require.NoError(t, NewPodBuilder("custom-dns-server", data.testNamespace, customDnsImage).
+		WithLabels(map[string]string{"app": "custom-dns"}).
+		WithContainerName("coredns").
+		WithArgs([]string{"-conf", "/etc/coredns/Corefile"}).
+		AddVolume(volume).AddVolumeMount(volumeMount).
+		WithAnnotations(map[string]string{randomPatchAnnotationKey: randSeq(annotationValueLen)}).
```

Comment:
nit: I think there is no need to have an annotation initially? That will be one less line of code...